lib/slurm_squeue_client.rb

```diff
@@ -61,7 +61,73 @@ def sinfo
     
     s.success? ? @sinfo = o : raise(CommandFailed, e)
   end
-  
+
+  # Return length of GRES field from SLURM
+  # @return [Integer] gres length
+  def gres_length
+    return @gres_length if definde?(@gres_length)
+
+    Open3.pipeline_rw "sinfo -o '%G' | awk '{ print length }'", "sort -n", "tail -1" do |stdin,stdout|
+      stdin.write stdout
+      stdin.close
+      @gres_length = stdout.read.to_i
+    end
+  end
+
+  # Parse and return total number of GPU nodes in a SLURM cluster
+  # @return [Integer] number of GPU nodes
+  def gpu_nodes
+    return @available_gpu_nodes if defined?(@available_gpu_nodes)
+
+    Open3.pipeline_rw "sinfo -N -h -a --Format='nodehost,gres:#{@gres_length} available'", 'uniq', 'grep gpu:v', 'wc -l' do |stdin, stdout|
+      stdin.write stdout
+      stdin.close
+      @available_gpu_nodes = stdout.read.to_i
+    end
+  end
+
+  # Return number of GPU nodes with mixed or idle status
+  # 
+  # @return [Integer] number of GPU nodes with status mixed (some CPUs allocated)
+  def gpu_nodes_free
+    return @gpu_nodes_free if defined?(@gpu_nodes_free)
+
+    Open3.pipeline_rw "sinfo -a -h --states=mixed,idle --Node --Format='nodehost,gres:#{@gres_length}'", 'uniq', 'grep gpu:v', 'wc -l' do |stdin, stdout|
```

Comment:
Remove `--states` and instead do `egrep -v 'mixed|alloc'`. Also a mixed node isn't necessarily idle, it's just not 100% allocated.